Better type inference, improved test coverage throughout locked dependencies, automated dependency upgrades

Ocramius opened the pull request:
This patch is mostly raised by the upgrade of `0.3.0` to `0.4.0` breaking BC (expected, but preventable).

In practice, Valinor is probably going to be used because of `TreeMapper#map(class-string<T>, mixed): T` in the grand majority of cases.

To cover this scenario, and prevent regressions, I introduced:

 * testing against **locked** dependencies, so that we can pinpoint tiny upgrades that break `cuyz/valinor` as they come up
 * automated dependency **widening** (I would prefer bumping, but that's really your decision)
 * `vimeo/psalm` type inference testing
 * better types around `TreeMapper#map()`, so that it works without coupling `cuyz/valinor` to a specific `vimeo/psalm` version (the type declarations for the majority scenario are in the code, without any need for plugins)

Ocramius commented on composer.json:
```diff
@@ -27,7 +27,8 @@
         "friendsofphp/php-cs-fixer": "^3.4",
-        "marcocesarato/php-conventional-changelog": "^1.12"
+        "marcocesarato/php-conventional-changelog": "^1.12",
+        "vimeo/psalm": "^4.18.1"
```
This was added on purpose, since we are verifying type inference in `tests/StaticAnalysis`

Ocramius commented on tests/StaticAnalysis/can-infer-tree-mapper-results.php:
```diff
@@ -0,0 +1,14 @@
+use CuyZ\Valinor\Tests\StaticAnalysis\Stub\A;
+
+/** @param mixed $input */
+function mappingClassWillInferObjectOfSameType(TreeMapper $mapper, $input): A
```
Unsure if we need docs for this one - perhaps `README.md` in the directory :thinking:

romm replied:
Nah it's pretty clear IMO.

romm replied:
Nah it's pretty clear IMO.

bendavies commented on .github/dependabot.yml:
```diff
@@ -0,0 +1,25 @@
+{
```
json in a `.yml` file?

Ocramius replied:
Yeah, makes it more manageable

bendavies replied:
TIL. didn't know this was supported.

romm commented on .github/dependabot.yml:
```diff
@@ -0,0 +1,27 @@
+                "interval": "weekly",
+                "time": "02:00"
+            },
+            "versioning-strategy": "widen"
```
I'm not very familiar to Dependabot, if I understand correctly [the doc](https://docs.github.com/en/code-security/supply-chain-security/keeping-your-dependencies-updated-automatically/configuration-options-for-dependency-updates#versioning-strategy) you're suggesting to use `increase` instead?

Also, how does it work by default? A pull request will automatically be opened to bump to new versions?

Ocramius replied:
Explained in https://github.com/CuyZ/Valinor/pull/59#issuecomment-1010922020

romm commented on .github/workflows/mutation.yml:
```diff
@@ -39,18 +39,7 @@ jobs:
-
-      - name: Install dependencies
-        run: composer update --prefer-dist --no-interaction
+      - uses: "ramsey/composer-install@v2"
```
Noice! Didn't know that one.

romm commented on src/Mapper/TreeMapper.php:
```diff
@@ -7,11 +7,18 @@
+     * @psalm-template TypeDefinition of string|class-string<TObject>
      *
-     * @param string|class-string<T> $signature
+     * @param TypeDefinition $signature
```
```suggestion
     * @param string|class-string<TObject> $signature
```
PhpStorm doesn't recognize the Psalm template and will not provide auto-completion if used. I'd prefer to keep this signature (unless you did it for a particular reason?). Also, the `@psalm-template TypeDefinition of string|class-string<TObject>` above should be removed.

Ocramius replied:
The point is giving PHPStan/Psalm more precise types, not to help PHPStorm.

If PHPStorm support is wished, I can provide a `.phpstorm.meta.php`, which does the auto-completion part: PHPStorm will eventually catch up with PHPStan/Psalm on this stuff.

Ocramius replied:
I'll see if I can simplify the signature first, BTW :+1:

romm replied:
> The point is giving PHPStan/Psalm more precise types, not to help PHPStorm.

Sure, I got that, it's just that I believe we can have both PHPStan/Psalm support **and** PhpStorm autocompletion (very convenient for daily usage IMO).

The signature `string|class-string<TObject>` is understood by PhpStorm, and it looks like it would be ok for PHPStan/Psalm as well. Tell me if I'm wrong :D

Ocramius replied:
I just need to see if I can use the conditional here :+1:

Ocramius replied:
0bf2a13b8feac683276d8930da428f93c10246f8 simplifies this: solved

romm commented on composer.lock:
```diff
@@ -203,31 +203,31 @@
         {
             "name": "psr/simple-cache",
-            "version": "2.0.0",
+            "version": "1.0.1",
```
Looks like there are a lot of changes, possible you were using PHP < 8.1? I think it would be best to always lock to dependencies with PHP 8.1 (where all features are enabled). WDYT?

Ocramius replied:
I did a `composer update` after setting `{"config": {"platform": {"php": "7.4.99"}}}` on purpose, so that the resulting `composer.lock` can then be tested on `7.4`, `8.0` and `8.1`.

romm replied:
Oh ok I see, this is related to the changes to the dependencies matrix in the tests pipeline. 👍 

I'm not used to the usage of `{"config": {"platform": {"php": "7.4.99"}}}` — I guess you didn't commit that change because the dependencies upgrades will be managed by Dependabot, right?

Ocramius replied:
Correct, I did:

 * manually add `{"config": {"platform": {"php": "7.4.99"}}}`
 * `composer update`
 * manually remove `{"config": {"platform": {"php": "7.4.99"}}}`
 * `composer update --lock`

That led to a PHP 7.4-compatible `composer.lock`

romm reviewed (commented):
Looking nice, thank you!

I'd like some explanation on the Dependabot thing, I often saw it on popular repos but I'm not sure how it works. 😅

Also, I try to keep using [conventional commits](https://www.conventionalcommits.org) for this repo (for consistency mainly). What would be your advice for this kind of PR where fixup commits were made? Should I squash it with a correct title/description when submitting the PR?

romm commented:
> Note: mutation tests fail because of no actual source diff in the patch.
> 
> Everything is done in:
> 
>     * `composer.json` + `composer.lock`
> 
>     * github actions manifests
> 
>     * dependabot manifest
> 
>     * docblocks

@Ocramius damn I don't get why it would work in #60 but not here. Any recommendation on how to fix this?

Ocramius commented:
> I'd like some explanation on the Dependabot thing, I often saw it on popular repos but I'm not sure how it works. sweat_smile

It will, once a week, look for new dependencies, `composer.json` and `composer.lock` accordingly (and also `.github/*` actions references), and send you PR for review for each dependency, so you don't need to chase changes in the ecosystem.

It will also rebase any PRs it currently has open, should `composer.json` and `composer.lock` change in between, so you don't need to do that work yourself.

 > Also, I try to keep using conventional commits for this repo (for consistency mainly).

I can squash out some commits, but no idea about what the actual commits here should be then? I keep them very much human-friendly, rather than "changed X", so if you want specific messages, please do tell.

 > What would be your advice for this kind of PR where fixup commits were made? Should I squash it with a correct title/description when submitting the PR?

Fixup commits will be squashed from my end, after review is through.

Ocramius commented:
Note: instead of a51f8c5 , I will rebase this patch post-review.

Merges of mainline into feature branches are really to be avoided (in pretty much every environment), and are a much bigger problem than conventional commits ;-)

Ocramius commented:
> > Note: mutation tests fail because of no actual source diff in the patch.
> > Everything is done in:
> > ```
> > * `composer.json` + `composer.lock`
> > 
> > * github actions manifests
> > 
> > * dependabot manifest
> > 
> > * docblocks
> > ```
> 
> @Ocramius damn I don't get why it would work in #60 but not here. Any recommendation on how to fix this?

This is probably because I have a diff in `tests/`

Ocramius commented:
> > I can squash out some commits, but no idea about what the actual commits here should be then? I keep them very much human-friendly, rather than "changed X", so if you want specific messages, please do tell.
> 
> The commits names are good, but what is missing is a correct prefix — see [conventional commits](https://www.conventionalcommits.org). This allows me to use [marcocesarato/php-conventional-changelog](https://github.com/marcocesarato/php-conventional-changelog) for the changelog automatic generation.

Rebased and added prefixes :muscle: 


> > Note: instead of [a51f8c5](https://github.com/CuyZ/Valinor/commit/a51f8c523bd07cc112ec278aaf83b5f70d1ce5cb) , I will rebase this patch post-review.
> > Merges of mainline into feature branches are really to be avoided (in pretty much every environment), and are a much bigger problem than conventional commits ;-)
> 
> Right, I usually use rebase strategies on my own branches, but I was not sure if that would fit on a PR managed by someone else. Learning basic OSS contribution rules here :D

Understandable: most people probably roll their eyes when I ask them for a rebase, but I do `git rebase` 10~20 times a day for my own stuff ;-)

> > This is probably because I have a diff in src/, but it doesn't involve any AST node changes, beside docblock
> 
> Ah! Right, must be that. Any idea how we could fix it, or avoid this case?

This is potentially a bug in `infection/infection` - I wouldn't worry too much about it, for now.

romm commented:
> Rebased and added prefixes 💪

Great! Not so important but FYI:
- For consistency the first letter of the subject is lowercase
- I force myself to have commit subjects that are less than 72 chars, this eases readability of the history — for instance GitHub will "break" subjects with more than 72 chars, this makes it harder to read:
  ![image](https://user-images.githubusercontent.com/6342901/149382399-3c0cc25f-b323-47b0-9a66-c7267bfa522f.png)

These are non-blocking from my POW but just wanted to share. 🙂 

> Understandable: most people probably roll their eyes when I ask them for a rebase, but I do `git rebase` 10~20 times a day for my own stuff ;-)

I also do, this is without doubt the best strategy to adopt, although rebasing on a contributor PR could also frighten the author if not used to it 😬 

> This is potentially a bug in `infection/infection` - I wouldn't worry too much about it, for now.

Alright, I'll see if I can report/fix it.

romm commented:
Just though I could easily reword the commits to fit the requirements I was writing about earlier — the sense doesn't change, it just fits the 72 chars threshold.  Hope you don't mind? I'm not sure if that can seem rude. 😬 

Anyway, thanks a lot for the contribution @Ocramius!

romm commented:
@Ocramius I'm not sure whether I should enable "Dependabot security updates" in the repository settings?

![image](https://user-images.githubusercontent.com/6342901/149392749-eaf3faba-a557-4720-884c-b0a083d685a9.png)

Ocramius commented:
Thanks @romm!

Enabling security updates shouldn't be necessary: weekly updates are enough for a library
